src/ucs/datastruct/ptr_map.h

```diff
@@ -0,0 +1,107 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "khash.h"
+#include <ucs/sys/compiler.h>
+#include <ucs/type/status.h>
+
+#include <stdint.h>
+
+#define UCS_PTR_MAP_KEY_INVALID    UINT64_MAX
+#define UCS_PTR_MAP_KEY_ID_FLAG    0x1
+
+
+/* Hash map to find pointer by its key */
+typedef uint64_t ucs_ptr_map_key_t;
+KHASH_TYPE(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*);
+typedef khash_t(ucs_ptr_map_impl) ucs_ptr_hash_t;
+KHASH_IMPL(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*, 1,
+           kh_int64_hash_func, kh_int64_hash_equal);
+
+
+typedef struct ucs_ptr_map {
+    uint64_t        cntr;
+    ucs_ptr_hash_t  hash;
+} ucs_ptr_map_t;
+
+
+ucs_status_t ucs_ptr_map_init(ucs_ptr_map_t *map);
+
+void ucs_ptr_map_destroy(ucs_ptr_map_t *map);
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key = (uintptr_t)ptr;
+
+    UCS_STATIC_ASSERT(UCS_PTR_MAP_KEY_INVALID & UCS_PTR_MAP_KEY_ID_FLAG);
+    ucs_assert_always(!(key & UCS_PTR_MAP_KEY_ID_FLAG));
+
+    return key;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put_safe(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key;
+    khiter_t iter;
+    int ret;
+
+    do {
+        /* wrap around counter */
+        do {
+            key = ((++map->cntr) << 1) | UCS_PTR_MAP_KEY_ID_FLAG;
+            if (ucs_unlikely(key == UCS_PTR_MAP_KEY_INVALID)) {
+                map->cntr = 0;
+                continue;
+            }
+        } while(0);
+
+        iter = kh_put(ucs_ptr_map_impl, &map->hash, key, &ret);
+        if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
+            return UCS_PTR_MAP_KEY_INVALID;
+        }
+    } while (ucs_unlikely(ret == UCS_KH_PUT_KEY_PRESENT));
+
+    kh_value(&map->hash, iter) = ptr;
+    return key;
+}
+
+static UCS_F_ALWAYS_INLINE void*
+ucs_ptr_map_get(const ucs_ptr_map_t *map, ucs_ptr_map_key_t key)
+{
+    khiter_t iter;
+
+    UCS_STATIC_ASSERT(UCS_PTR_MAP_KEY_INVALID & UCS_PTR_MAP_KEY_ID_FLAG);
+
+    if (key & UCS_PTR_MAP_KEY_ID_FLAG) {
+        /* UCS_PTR_MAP_KEY_INVALID is never put, so error input can be checked
+         * here to avoid extra branch */
+        iter = kh_get(ucs_ptr_map_impl, &map->hash, key);
+        return ucs_unlikely(iter == kh_end(&map->hash)) ? NULL :
+               kh_value(&map->hash, iter);
+    } else {
```

Comment:
```else``` is not needed

Reply:
if/else is used because these cases are equal in possibility

Reply:
maybe the case where the key is the pointer itself can be considered "fast path"? (e.g no err handling)

Reply:
ok